### ISSUE TITLE
Needs some general  help with Transporter/Cargo

### ISSUE THREAD

**Issue by Iamblichos** [LoneGazebo/Community-Patch-DLL#1344]
I have an interesting condition where I have a land unit(s) loaded on a Melee boat and those land units have an attack target. On one hand (1), if the transporter is full or cargospaceavailable is 0 than any land melee unit on board cannot attack, on the other hand (2) if there is cargospaceavailable still, than any land melee unit on board can do the attack move. Non-issue with ranged units since they aren't leaving and entering a plot:

(1):
![2016-02-28_00001](https://cloud.githubusercontent.com/assets/14372214/13380822/3a817f84-de11-11e5-86cf-c6b536309dfb.jpg)

(2):

![2016-02-28_00002](https://cloud.githubusercontent.com/assets/14372214/13380823/44bbd094-de11-11e5-9b26-128de52b245f.jpg)

I know this has to do with cargospaceavailable on the ship and for some reason it treats the land unit leaving to perform the attack as adding to the cargocount since it will have to return to the ship. I tried just commenting out in bool CvUnit::canLoadUnit(const CvUnit& unit, const CvPlot& targetPlot) const:

```
if(!(unit.cargoSpaceAvailable(getSpecialUnitType(), getDomainType())))
	{
		return false;
	}
```
...to see if the one that is full could attack, and sure enough they could, but then doing that would allow an infinite amount of units to load into ship. Racked my brain over this for a  couple of days and haven't made any head way. If anybody has any insight, could use the help.

**Comment by LoneGazebo** [comment]
Why not just eliminate ability of transport units to attack?

**Comment by Iamblichos** [comment]
I could but that would go against my intention. With the way I've slightly edited the transporter with land cargo, it's treated the cargo is in Native Domain so it ought to be able to attack from the boat if in proximity of a valid move. I thought about nulling out the transporter if the cargo unit is attacking but that might be a bad idea to totally dislodge it from the transporter.

**Comment by Iamblichos** [comment]
The transporter is coded to take into account ranged unit, so this is an unknown condition to the code as far as I can tell.

**Comment by Iamblichos** [comment]
Another idea I have is to subtract 1 count in cargospaceavailable if canattackwithmove && isattacking...

**Comment by Iamblichos** [comment]
came up with a hack and it works but it can be manipulated if you have two adjacent boats with cargo and both have cargo that can attack the plot:

```
int CvUnit::cargoSpaceAvailable(SpecialUnitTypes eSpecialCargo, DomainTypes eDomainCargo) const
{
	VALIDATE_OBJECT
	IDInfo* pUnitNode;
	CvUnit* pLoopUnit;
	CvPlot* pPlot;
	if(specialCargo() != NO_SPECIALUNIT)
	{
		if(specialCargo() != eSpecialCargo)
		{
			return 0;
		}
	}
	if(domainCargo() != NO_DOMAIN)
	{
		if(domainCargo() != eDomainCargo)
		{
			return 0;
		}
	}
	if(domainCargo() == DOMAIN_LAND)
	{
		pPlot = plot();
		pUnitNode = pPlot->headUnitNode();
		while (pUnitNode != NULL)
		{
			pLoopUnit = ::getUnit(*pUnitNode);
			pUnitNode = pPlot->nextUnitNode(pUnitNode);
			if(pLoopUnit)
			{
				CvPlot* pAdjacentPlot;
				int iI;
				for(iI = 0; iI < NUM_DIRECTION_TYPES; ++iI)
				{
					pAdjacentPlot = plotDirection(getX(), getY(), ((DirectionTypes)iI));
					if(pAdjacentPlot != NULL)
					{
						if(pLoopUnit->isCargo() && pLoopUnit->IsCanAttackWithMove() && pLoopUnit->canMoveInto(*pAdjacentPlot, MOVEFLAG_ATTACK))
						{
							return std::max(0, (cargoSpace() - getCargo() + 1));
						}
					}
				}
			}
		}
		return std::max(0, (cargoSpace() - getCargo()));
	}
	else
	{
		return std::max(0, (cargoSpace() - getCargo()));
	}
}
```

Tried a thoudsand different other things with no success.

**Comment by LoneGazebo** [comment]
I'd ask @ilteroi , this is his kind of thing, not mine. I'm more of an AI man myself. :)

**Comment by Iamblichos** [comment]
paging the good doctor!

**Comment by ilteroi** [comment]
somebody mentioned my name?

i think the root cause is that movement for melee units is not guaranteed to succeed. because there might be an enemy unit in the plot and it's not guaranteed that it can be destroyed (regardless of odds). therefore the pathfinder only allows attacks if the unit can stay in it's current place.

now in the pathfinder it says:
```
	kToNodeCacheData.bFriendlyUnitLimitReached = (pPlot->getNumFriendlyUnitsOfType(pUnit) >= pPlot->getUnitLimit());
```

and especially getNumFriendlyUnitsOfType() looks like it might fail with unexpected embark states.

**Comment by Iamblichos** [comment]
Yeah I see that, but since you aren't changing the transport state from say 2 to 3, (i.e. you aren't adding a new unit), it falls back to it's position in the cargo slot, and ...getNumFriendlyUnitsofType() should always be 0 for cargo units.

**Comment by ilteroi** [comment]
come to think of it, i'm wondering why it says >= in the line above. wouldn't > be more correct?

**Comment by Iamblichos** [comment]
yeah it seems that way...unitlimit should only break when you are > getUnitLimit()

**Comment by Iamblichos** [comment]
changing that line from >=  to > didn't resolve it.

**Comment by Iamblichos** [comment]
There's also this issue in pathfinder:
![2016-02-29_00001](https://cloud.githubusercontent.com/assets/14372214/13410385/1e4910a6-defd-11e5-8ccc-ae9369ed01d7.jpg)
The first movement into the available slotted melee boat should absorb all movement cost.

**Comment by ilteroi** [comment]
well, if it's not embarkation then all bets are off ... and i feel i have to warn you: the movement rules in this game are an absolute mess, poorly thought out and poorly implemented. be careful when changing anything, i'm still fixing bugs from my last attempt.

**Comment by Iamblichos** [comment]
I may just stick with the hack above to fool the cargospaceavailable until we find a solution.

**Comment by Iamblichos** [comment]
@literoi just to clarify one point, I've only made a couple of edits in CvPlot to this and they seem to behave fine and safe edits, and I did it for the AI behavior. Land units treat available adjacent-to-land Transporter ships (i.e. Naval Melee Ships) as a part of their native domain. Transporter ships will escort them to a destination target. I also edited DLL so that embarked land units can't load, otherwise it'll confuse their movement path if a melee ship falls within the path of embarked unit. Movement seems all fine and consistent on that end.
